### DISCUSSION TITLE
How can I pre-render data (server side) which requires authentication?

### DISCUSSION THREAD

**Discussion by ben519** [vercel/next.js#50669] [App Router] [answered]
I'm trying reproduce something similar to LinkedIn. It goes like this..

1. I have a Firestore collection of *user* documents. Each user has a `name` and an `email`.

    <img width="994" alt="Screenshot 2023-06-01 at 8 59 36 PM" src="https://github.com/vercel/next.js/assets/723099/d429c74e-e384-497d-9d6b-3bf4c70d846e">

2. I want to generate a profile page for each user at the route `user/[id]`
    1. Authenticated users should see the user's name and email.
        
        <img width="387" alt="Screenshot 2023-06-01 at 9 11 34 PM" src="https://github.com/vercel/next.js/assets/723099/0e0bf573-218a-44f6-9f35-d7deac5936bb">

    2. Unauthenticated users (including web crawlers) should see the user's name when they visit a user's profile.
         
        <img width="392" alt="Screenshot 2023-06-01 at 9 11 28 PM" src="https://github.com/vercel/next.js/assets/723099/b0c0fc3f-a96c-46f6-934a-52d19e63f9a8">

    **The reason this is so important to me is SEO**. I want Google to find and index these profile pages.

**Important Notes**  
- You cannot request part of a Firestore document. You can either fetch all of it or none of it.
- My Firestore rules prevent you from fetching a user document unless you are authenticated. The rules look like this

    ```
    rules_version = '2';
    
    service cloud.firestore {
      match /databases/{database}/documents {
        match /{document=**} {
    	    match /users/{id} {
      	    allow read: if request.auth.uid != null;
    	    }
        }
      }
    }
    ```

---

Is this pattern possible to achieve? Assuming it is, how can I go about it? Any examples would be great.

**Comment by SamKomesarook** [comment, marked as answer]
If I understand your goal correctly, maybe you could create a server page that utilizes [generateStaticParams](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes#generating-static-params), and references the data that requires auth in a client component like the below:

```typescript
export async function generateStaticParams() {
  const users = await getUsers() // Get users from 

  return users.map((user: any) => ({
    name: user.name,
  }))
}
export default function AccountPage({ params }: { params: { name: string } }) {
  return (
    <div>
      {name}
<AuthenticatedCheck />
    </div>
  )
}
``` 

```typescript
"use client"

export default function AuthenticatedCheck({name: string}) {
  return (
    <div>
    {authenticated ? <AuthenticatedInfo name={name}/> : <></>}
    </div>
  )
}
``` 

```typescript
"use client"

const getEmail = (name) => ... // Hook to get email 

export default function AuthenticatedInfo({name: string}) {
const email = use(getEmail(name))
  return (
   <div>{email}</div>
  )
}
``` 

I simplified a lot in the example, but the idea is the server component of just the username will be built server-side, then when the client authenticates themselves, they will see the email addr in a client component.

As for the firestore notes: 
1. There would be a bit of redundancy with your requests if you can only request all or none of the data, but I'm not sure how big of an issue that is for your setup...
2. You would have to make sure the call in the server component's getStaticProps was authenticated properly, and likewise for the client components, obviously...

**Reply by ben519** [reply]
Hey thanks. This is the direction I was headed. [This article](https://colinhacks.com/essays/nextjs-firebase-authentication) was quite helpful as well.

The thing that annoys me is that this pattern causes a slight delay in fetching & rendering data for my authenticated users. But I can live with that for now.

**Reply by SamKomesarook** [reply]
You could wrap the client component in a [suspense boundary](https://nextjs.org/docs/app/building-your-application/routing/loading-ui-and-streaming#example), so at least some loading state was provided in the meantime.

**Reply by ben519** [reply]
Good idea. I'll give this a shot. Thanks!
